drv/transceivers-server/src/net_api.rs

```diff
@@ -0,0 +1,410 @@
+                len: mem.len(),
+            });
+        }
+        // CMIS spec only guarantee an 8-byte read (?!)
+        // (see CMIS 5.0, 5.2.2.1)
+        if matches!(mem.upper_page(), UpperPage::Cmis(..)) && mem.len() > 8 {
+            // TODO: more specialized error
+            return Err(Error::RequestTooLarge);
+        }
+
+        self.select_page(*mem.upper_page(), modules)?;
+        self.transceivers
+            .setup_i2c_read(mem.offset(), mem.len(), get_mask(modules)?)
+            .map_err(|_e| Error::ReadFailed)?;
+        sleep_for(5); // TODO: wait for completion
+
+        let controller = match modules.fpga_id {
+            0 => FpgaController::Left,
+            1 => FpgaController::Right,
+            i => return Err(Error::InvalidFpga(i)),
+        };
+        for (port, out) in modules
+            .ports
+            .to_indices()
+            .zip(out.chunks_mut(mem.len() as usize))
+        {
+            self.transceivers
+                .get_i2c_read_buffer(PortLocation { controller, port }, out)
+                .map_err(|_e| Error::ReadFailed)?;
+        }
+        Ok(())
+    }
+
+    fn write(
+        &mut self,
+        mem: MemoryRegion,
+        modules: ModuleId,
+        data: &[u8],
+    ) -> Result<(), Error> {
+        // Writing beyond a 128-byte page boundary will wrap around; return an
+        // error if that's going to happen.
+        let page_end = (mem.offset() as u32 / 128 + 1) * 128;
+        if mem.offset() as u32 + mem.len() as u32 > page_end {
+            // TODO use a more descriptive error name here?
+            return Err(Error::InvalidMemoryAccess {
+                offset: mem.offset(),
+                len: mem.len(),
+            });
+        }
+        match *mem.upper_page() {
+            // CMIS 5.0, §5.2.2.2: only guarantee an 8-byte write
+            UpperPage::Cmis(..) => {
+                if mem.len() > 8 {
+                    // TODO: more specialized error
+                    return Err(Error::RequestTooLarge);
+                }
+            }
+            // SFF-8636, §5.3.3: max of 4 bytes in one write transaction
+            UpperPage::Sff8636(..) => {
+                if mem.len() > 4 {
+                    // TODO: more specialized error
+                    return Err(Error::RequestTooLarge);
+                }
+            }
+        }
+
+        self.select_page(*mem.upper_page(), modules)
+            .map_err(|_e| Error::WriteFailed)?;
+
+        // Copy data into the FPGA write buffer
+        self.transceivers
+            .set_i2c_write_buffer(&data[..mem.len() as usize])
+            .map_err(|_e| Error::WriteFailed)?;
+
+        // Trigger a multicast write to all transceivers in the mask
+        self.transceivers
+            .setup_i2c_write(mem.offset(), mem.len(), get_mask(modules)?)
+            .map_err(|_e| Error::WriteFailed)?;
+        sleep_for(5); // TODO: wait for completion
+
+        Ok(())
+    }
+}
```

Comment:
As I noted in a few places above, as of https://github.com/oxidecomputer/hubris/commit/ae111d6b5a027c1c987cc7ce8e8eb5c663cb82ca we have registers to poll for I2C bus busy per port. Additionally, there are error registers we should also be checking to make sure the transaction has not only completed, but completed successfully.